cbmc/proofs/IotMqtt_Connect/IotMqtt_Connect_harness.c

```diff
@@ -0,0 +1,140 @@
+#include "private/iot_mqtt_internal.h"
+#include "iot_mqtt_protocol.h"
+
+#include <stdlib.h>
+#include <string.h>
+
+#include "mqtt_state.h"
+
+typedef bool ( * MatchFunction_t )( const IotLink_t * const pOperationLink,
+                                    void * pCompare );
+
+typedef void ( * FreeElementFunction_t )( void * pData );
+
+/**
+ * We constrain the return values of this function because it
+ * is checked by assertions in the MQTT code.
+ */
+IotTaskPoolError_t IotTaskPool_TryCancel( IotTaskPool_t taskPool,
+                                          IotTaskPoolJob_t job,
+                                          IotTaskPoolJobStatus_t * const pStatus )
+{
+    assert( ( taskPool == NULL ) || ( job == NULL ) );
+    return IOT_TASKPOOL_BAD_PARAMETER;
+}
+
+/**
+ * We constrain the return values of these functions because
+ * they are checked by assertions in the MQTT code.
+ */
+IotTaskPoolError_t IotTaskPool_CreateJob( IotTaskPoolRoutine_t userCallback,
+                                          void * pUserContext,
+                                          IotTaskPoolJobStorage_t * const pJobStorage,
+                                          IotTaskPoolJob_t * const pJob )
+{
+    assert( userCallback != NULL );
+    assert( pJobStorage != NULL );
+    assert( pJob != NULL );
+
+    /* _IotMqtt_ScheduleOperation asserts this. */
+    return IOT_TASKPOOL_SUCCESS;
+}
+
+/**
+ * We abstract these functions for performance reasons. In their original
+ * implementation, CBMC ended up creating byte extracts for all possible
+ * objects, due to the polymorphic nature of the linked list. We assume
+ * that the functions are memory safe, and we free and havoc the list to
+ * demonstrate that no subsequent code makes use of the values in the list.
+ */
+void IotListDouble_RemoveAllMatches( const IotListDouble_t * const pList,
+                                     MatchFunction_t isMatch,
+                                     void * pMatch,
+                                     FreeElementFunction_t freeElement,
+                                     size_t linkOffset )
+{
+    free_IotMqttSubscriptionList( pList );
+    allocate_IotMqttSubscriptionList( pList, SUBSCRIPTION_COUNT_MAX - 1 );
+}
+
+void _IotMqtt_RemoveSubscriptionByTopicFilter( _mqttConnection_t * pMqttConnection,
+                                               const IotMqttSubscription_t * pSubscriptionList,
+                                               size_t subscriptionCount )
+{
+    free_IotMqttSubscriptionList( &( pMqttConnection->subscriptionList ) );
+    allocate_IotMqttSubscriptionList( &( pMqttConnection->subscriptionList ), subscriptionCount );
+}
+
+/**
+ * We abstract the concurrency related functions because this is a sequential proof.
+ * But we implement the following stubs to improve coverage.
+ */
+bool IotSemaphore_TimedWait( IotSemaphore_t * pSemaphore,
+                             uint32_t timeoutMs )
+{
+    assert( pSemaphore != NULL );
```

Comment:
This is kind of unnecessary, it is an assert, but the result is what you would get by default by leaving the function out of the goto binary.